### PR TITLE
feat(dev): Stop code generation action

### PR DESCRIPTION
**Pull request by tverney**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description
Currently /dev don't support stop code generation. This PR introduces this functionality (watch the video below).

Mynah UI provides onStopChatResponse API which we can hook in the cancellation token provided on VS Code, sharing across an active session, aborting current progress.

https://github.com/user-attachments/assets/0ab591f3-aed1-41e4-a5a9-5e405a404c3b

VS Code side: https://github.com/aws/aws-toolkit-vscode/pull/5675

## Checklist
- [X] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [X] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [X] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererConstants.kt
Comment (tverney) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererConstants.kt:122]: Good catch, it was supposed to be ignored in the first commit.

**Review by neilk-aws — APPROVED** [review]

**Comment by tverney** [comment]
!!Important - FYI toolkit team: this is only to be merged in next week's release. CC: @neilk-aws

**Review by neilk-aws — CHANGES_REQUESTED** [review]
Blocking to wait for security sign-off

**Comment by tverney** [comment]
We have already the security sign-off. Ready to be reviewed cc: @neilk-aws

**Review by neilk-aws — APPROVED** [review]

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevControllerExtensions.kt
Comment (rli) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevControllerExtensions.kt:163]: why wouldnt these always be `Int`? why do we allow this to be nullable?

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevControllerExtensions.kt
Comment (rli) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevControllerExtensions.kt:39]:
```suggestion
        message = if (session.retries == CODE_GENERATION_RETRY_LIMIT) {
```

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevControllerExtensions.kt
Comment (rli) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevControllerExtensions.kt:158]: see code scanner comment. why does everything need to be an extension?

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/CodeGenerationState.kt
Comment (rli) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/CodeGenerationState.kt:36]: still unsure why we're allowing null here

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevController.kt
Comment (rli) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevController.kt:292]:
```suggestion
```
declared too early.
var can be val

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/CodeGenerationState.kt
Comment (rli) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/CodeGenerationState.kt:42]: why are these in the constructor? aren't they inherent properties of the state?

**Review by rli — COMMENTED** [review]
very difficult to review with all the whitespace changes

File: detekt-rules/tst/software/aws/toolkits/gradle/detekt/rules/LazyLogRuleTest.kt
Comment (tverney) [reply detekt-rules/tst/software/aws/toolkits/gradle/detekt/rules/LazyLogRuleTest.kt:1]: Build was failing because of this, not related to the change. I'll revert, but just FYI.

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/Session.kt
Comment (tverney) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/Session.kt:85]: I don't see any other interface for cancellation token shared in this project besides this one. I'm reusing the same for IDE extension (IdeExtension.kt:81). Please, indicate if there is any recommended by the team.

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevControllerExtensions.kt
Comment (tverney) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevControllerExtensions.kt]: Removed, we can use as a private function only, not attached to the extension

File: plugins/core/jetbrains-community/resources/telemetryOverride.json
Comment (tverney) [reply plugins/core/jetbrains-community/resources/telemetryOverride.json:1]: The metric changes are made by @kelvin-klchu, he will follow up there.

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/CodeGenerationState.kt
Comment (tverney) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/CodeGenerationState.kt:42]: Its inherit, but also modified by the the current context, so need be overrided.

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevController.kt
Comment (tverney) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevController.kt:292]: Value is reassigned on line 305, so need to be var. I've removed the redundant null

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/CodeGenerationState.kt
Comment (tverney) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/CodeGenerationState.kt:36]: Var-property type is Int, which is not a type of overridden

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevControllerExtensions.kt
Comment (tverney) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevControllerExtensions.kt:56]: Its reassigned afterwards in a condition, so need to be var

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevControllerExtensions.kt
Comment (tverney) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevControllerExtensions.kt]: Updated, thanks

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/SessionStateTypes.kt
Comment (tverney) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/SessionStateTypes.kt:7]: I think if they don't have gradle plugin installed it can't even run currently

Screenshot 2024-10-23 at 8 03 35 AM

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/Session.kt
Comment (rli) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/Session.kt:85]: `IdeExtension` is a util for telling the buildscript how to run certain tests.
`EmptyProgressIndicator` is closest in semantics to what you seem to be trying to do, but it appears that JetBrains has marked that as deprecated

Comment (tverney) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/Session.kt:85]: With Coroutines we will need wrap pretty much the whole feature dev to make it stateful. The CancellationToken its an approach that we can make it an event-based trigger fashion. 

The same its being used there https://github.com/aws/aws-toolkit-vscode/pull/5675

Its interesting, pretty much all the E2E tests are passing which suggests that works, when bundled. So what we're questioning here is just add a 3P tooling in feature dev or not?

Comment (tverney) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/Session.kt:85]: @rli Per sync with the team, we will try to implement our own CancellationToken and test its own state removing the Gradle dependency. We will providing in the next hours, I'll sync back here when we push it through.

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/SessionStateTypes.kt
Comment (tverney) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/SessionStateTypes.kt]: Removed gradle dependency

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/Session.kt
Comment (tverney) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/Session.kt]: @rli we've removed gradle dependency and added our own token.

File: plugins/amazonq/codetransform/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/BuildStepStatus.kt
Comment (rli) [inline plugins/amazonq/codetransform/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/BuildStepStatus.kt:20]: this will throw an exception

File: plugins/amazonq/mynah-ui/src/mynah-ui/ui/connector.ts
Comment (rli) [inline plugins/amazonq/mynah-ui/src/mynah-ui/ui/connector.ts:406]: was this intentional

Comment (tverney) [reply plugins/amazonq/mynah-ui/src/mynah-ui/ui/connector.ts]: added back to you :)

**Review by rli — APPROVED** [review]

**Pull request merged**
